contracts/external/cw-payroll/src/state.rs

```diff
@@ -0,0 +1,122 @@
+
+use crate::{balance::WrappedBalance, msg::StreamId};
+
+#[cw_serde]
+pub struct Config {
+    pub admin: Addr,
+}
+
+pub const CONFIG: Item<Config> = Item::new("config");
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct Stream {
+    pub admin: Addr,
+    pub recipient: Addr,
+    /// Balance in Native and Cw20 tokens
+    pub balance: WrappedBalance,
+    pub claimed_balance: WrappedBalance,
+    pub start_time: u64,
+    pub end_time: u64,
+    pub paused_time: Option<u64>,
+    pub paused_duration: Option<u64>,
+    pub paused: bool,
+
+    // METADATA
+    /// Title of the payroll item, for example for a bug bounty "Fix issue in contract.rs"
+    pub title: Option<String>,
+    /// Description of the payroll item, a more in depth description of how to meet the payroll conditions
+    pub description: Option<String>,
+    /// Link to stream attached for sync
+    pub link_id: Option<StreamId>,
+    /// Making a stream detachable will only affect linked streams.
+    /// A linked stream that detaches in the future will pause both streams.
+    /// Each stream must then resume on their own, or be fully removed to re-link.
+    pub is_detachable: bool,
+}
+
+impl Stream {
+    pub(crate) fn can_distribute_more(&self) -> bool {
+        if self.balance.amount() == 0 {
+            return false;
+        }
+        self.claimed_balance.amount() < self.balance.amount()
+    }
+    pub(crate) fn calc_distribution_rate_core(
+        start_time: u64,
+        end_time: u64,
+        block_time: Timestamp,
+        paused_duration: Option<u64>,
+        balance: &WrappedBalance,
+        claimed: &WrappedBalance,
+    ) -> (Uint128, Uint128) {
+        let block_time = std::cmp::min(block_time.seconds(), end_time);
+        let duration: u64 = end_time.checked_sub(start_time).unwrap_or_default();
+        if duration > 0 {
+            let diff = block_time.checked_sub(start_time).unwrap_or_default();
+
+            let passed: u128 = diff
+                .checked_sub(paused_duration.unwrap_or_default())
+                .unwrap_or_default()
+                .into();
+
+            let rate_per_second = balance
+                .amount()
+                .checked_div(duration.into())
+                .unwrap_or_default();
+
+            return (
+                Uint128::from(
+                    (passed * rate_per_second)
+                        .checked_sub(claimed.amount())
+                        .unwrap_or_default(),
+                ),
+                Uint128::from(rate_per_second),
+            );
+        }
+        (Uint128::new(0), Uint128::new(0))
+    }
+    pub(crate) fn calc_distribution_rate(&self, block_time: Timestamp) -> (Uint128, Uint128) {
+        Stream::calc_distribution_rate_core(
+            self.start_time,
+            self.end_time,
+            block_time,
+            self.paused_duration,
+            &self.balance,
+            &self.claimed_balance,
+        )
```

Comment:
why have this layer of indirection?

Reply:
Some basic calculations moved out of the contract, so the stream rate calculation logic can be used somewhere else